polaris/evaluate/utils.py

```diff
@@ -0,0 +1,122 @@
+import numpy as np
+import pandas as pd
+from typing import Union, Optional
+
+from polaris.evaluate import BenchmarkResults, ResultsType
+from polaris.utils.types import PredictionsType
+from polaris.evaluate import Metric
+
+def is_multi_task_single_test_set(vals: PredictionsType, target_cols: list[str]):
+    """Check if the given values are for a multiple-task benchmark with a single
+    test set. This is inferred by comparing the target names with the keys of the
+    given data. If all keys in the given data match the target column names, we
+    assume they are target names (as opposed to test set names for a single-task,
+    multiple test set benchmark)."""
+    return all(k in target_cols for k in vals)
+
+def serialize_predictions(
+    predictions: PredictionsType,
+):
+    """Used to serialize a Predictions object such that it can be sent over the wire during
+    external evaluation for competitions"""
+
+    if isinstance(predictions, np.ndarray):
+        return predictions.tolist()
+    elif isinstance(predictions, list):
+        return predictions
+    elif isinstance(predictions, dict):
+        for key, val in predictions.items():
+            predictions[key] = serialize_predictions(val)
+        return predictions
+    else:
+        raise TypeError(f"Unexpected type of {type(predictions)} encountered in Predictions object.")
+
+
+def deserialize_predictions(
+    predictions: PredictionsType,
+):
+    """Used to deserialize a Predictions object such that it can be converted to the
+    expected type of numpy array during external evaluation for competitions"""
+
+    if isinstance(predictions, np.ndarray):
+        return predictions
+    elif isinstance(predictions, list):
+        return np.array(predictions)
+    elif isinstance(predictions, dict):
+        for key, val in predictions.items():
+            predictions[key] = deserialize_predictions(val)
+        return predictions
+    else:
+        raise TypeError(f"Unexpected type of {type(predictions)} encountered in Predictions object.")
+
+
+def normalize_predictions_type(vals: PredictionsType, target_cols: list[str]):
+    if isinstance(vals, dict):
+        if is_multi_task_single_test_set(vals, target_cols):
+            return {"test": vals}
+        else:
+            return vals
+    elif vals is None:
+        return None
+    else:
+        return {"test": {target_cols[0]: vals}}
+
+def safe_mask(dict, test_label, target_label, mask):
+    if dict is None or dict[test_label] is None or dict[test_label][target_label] is None:
+        return None
+    else:
+        return dict[test_label][target_label][mask]
+
+def evaluate_benchmark(
```

Comment:
This is the shared evaluation logic. Regular and competition benchmarks use the exact same evaluation logic, but there has been some refactoring to facilitate sharing this code.